subsys/net/lib/azure_iot_hub/src/azure_iot_hub.c

```diff
@@ -0,0 +1,1348 @@
+/*
+ * Copyright (client) 2020 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <net/mqtt.h>
+#include <net/socket.h>
+#include <stdio.h>
+#include <net/azure_iot_hub.h>
+#include <settings/settings.h>
+
+#include "azure_iot_hub_dps.h"
+
+#if defined(CONFIG_AZURE_FOTA)
+#include <net/azure_fota.h>
+#endif
+
+#if defined(CONFIG_BOARD_QEMU_X86)
+#include "certificates.h"
+#endif
+
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(azure_iot_hub, CONFIG_AZURE_IOT_HUB_LOG_LEVEL);
+#define AZURE_IOT_HUB_AF_FAMILY AF_INET
+
+#define USER_NAME_STATIC	CONFIG_AZURE_IOT_HUB_HOSTNAME "/"	\
+				CONFIG_AZURE_IOT_HUB_DEVICE_ID		\
+				"/?api-version=2018-06-30"
+/* User name when connecting to Azure IoT hub is on the form
+ *	<IoT Hub hostname>/<device ID>/?api-version=2018-06-30
+ */
+#define USER_NAME_TEMPLATE	"%s/%s/?api-version=2018-06-30"
+
+#define TOPIC_DEVICEBOUND	"devices/%s/messages/devicebound/#"
+#define TOPIC_TWIN_DESIRED	"$iothub/twin/PATCH/properties/desired/#"
+#define TOPIC_TWIN_RES		"$iothub/twin/res/#"
+#define TOPIC_TWIN_REPORT	"$iothub/twin/PATCH/properties/reported/?$rid=%d"
+#define TOPIC_EVENTS		"devices/%s/messages/events/"
+#define TOPIC_TWIN_REQUEST	"$iothub/twin/GET/?$rid=%d"
+#define TOPIC_DIRECT_METHODS	"$iothub/methods/POST/#"
+#define TOPIC_DIRECT_METHOD_RES	"$iothub/methods/res/%d/?$rid=%d"
+
+#define DPS_USER_NAME		CONFIG_AZURE_IOT_HUB_DPS_ID_SCOPE \
+				"/registrations/%s/api-version=2019-03-31"
+
+#define FOTA_PROGRESS_REPORT	"{\"firmware\":{\"status\":\"%s\",\"dl\":%d}}"
```

Comment:
Could we add the `fwVersion` property here, which is copied from the same property on the desired object? Otherwise it is not clear *which* firmware the device is installing. Because it is not guaranteed that the device always picks up the latest FOTA job.

Reply:
Agreed. The current reporting mechanism works more like a placeholder for testing purposes.
The firmware version currently being downloaded should definitely be reported. It should also report back after the device has rebooted and actually applied the new firmware image to indicate whether that succeeded or not, ie implement the rest of the [recommended ](https://docs.microsoft.com/en-us/azure/iot-hub/tutorial-firmware-update) `fwUpdateStatus` states, along with `currentFwVersion`, `pendingFwVersion` and possibly `fwUpdateSubstatus`.

Reply:
To my understanding this is already solved:
1. when downloading the firmware, the device reports the **new** version in `pendingFwVersion` before rebooting
2. after reboot 
   1. if the update was **successful** the application will report its current version (which is hardcoded in the firmware) in `currentFwVersion` AND `pendingFwVersion`; they will now contain the *new* version.
   2. if the update was **not successful** `currentFwVersion` AND `pendingFwVersion` will then contain the *old* version.